packages/schematics/package-update/utility/npm.ts

```diff
@@ -0,0 +1,266 @@
+    } else {
+      return (loose ? '~' : '') + maybeMatch;
+    }
+  }
+}
+
+/**
+ * Get the NPM repository's package.json for a package. This is p
+ * @param {string} packageName The package name to fetch.
+ * @param {LoggerApi} logger A logger instance to log debug information.
+ * @returns {Observable<JsonObject>} An observable that will put the pacakge.json content.
+ * @private
+ */
+function _getNpmPackageJson(
+  packageName: string,
+  logger: logging.LoggerApi,
+): Observable<JsonObject> {
+  const url = `http://registry.npmjs.org/${packageName.replace(/\//g, '%2F')}`;
+  logger.debug(`Getting package.json from ${JSON.stringify(packageName)}...`);
+
+  let maybeRequest = npmPackageJsonCache.get(url);
+  if (!maybeRequest) {
+    const subject = new ReplaySubject<JsonObject>(1);
+
+    const request = http.request(url, response => {
+      let data = '';
+      response.on('data', chunk => data += chunk);
+      response.on('end', () => {
+        try {
+          const json = JSON.parse(data);
+          subject.next(json as JsonObject);
+          subject.complete();
+        } catch (err) {
+          subject.error(err);
+        }
+      });
+      response.on('error', err => subject.error(err));
+    });
+    request.end();
+
+    maybeRequest = subject.asObservable();
+    npmPackageJsonCache.set(url, maybeRequest);
+  }
+
+  return maybeRequest;
+}
+
+/**
+ * Recursively get versions of packages to update to, along with peer dependencies. Only recurse
+ * peer dependencies and only update versions of packages that are in the original package.json.
+ * @param {JsonObject} packageJson The original package.json to update.
+ * @param {{[p: string]: string}} packages
+ * @param {{[p: string]: string}} allVersions
+ * @param {LoggerApi} logger
+ * @param {boolean} loose
+ * @returns {Observable<void>}
+ * @private
+ */
+function _getRecursiveVersions(
+  packageJson: JsonObject,
+  packages: { [name: string]: string },
+  allVersions: { [name: string]: string },
+  logger: logging.LoggerApi,
+  loose: boolean,
+): Observable<void> {
+  return Observable.from(kPackageJsonDependencyFields).pipe(
+    mergeMap(field => {
+      const deps = packageJson[field] as JsonObject;
+      if (deps) {
+        return Observable.from(
+          Object.keys(deps)
+            .map(depName => depName in deps ? [depName, deps[depName]] : null)
+            .filter(x => !!x),
+        );
+      } else {
+        return Observable.empty();
+      }
+    }),
+    mergeMap(([depName, depVersion]: [string, string]) => {
+      if (!packages[depName] || packages[depName] === depVersion) {
+        return Observable.empty();
+      }
+      if (allVersions[depName] && semver.intersects(allVersions[depName], depVersion)) {
+        allVersions[depName] = semverIntersect.intersect(allVersions[depName], depVersion);
+
+        return Observable.empty();
+      }
+
+      return _getNpmPackageJson(depName, logger).pipe(
```

Comment:
Should this be cached somewhere to prevent recursive calls from making a lot of repeat http calls? I'm not sure how deep this recursion goes.

Reply:
It is. See `npmPackageJsonCache`.